app/services/school_data_importer.rb

```diff
@@ -87,4 +89,29 @@ def row_to_lad(row)
     local_authority_district.save!
     local_authority_district
   end
+
+  def link_school_to_local_authority(school, local_authority)
+    school_local_authority = SchoolLocalAuthority.find_or_initialize_by(school: school, local_authority: local_authority)
+
+    if school_local_authority.new_record?
+      school_local_authority.start_year = start_year
+      SchoolLocalAuthority.find_by(school: school, end_year: nil)&.update!(end_year: start_year)
+    end
+
+    school_local_authority.save!
+  end
+
+  def link_school_to_lad(school, lad)
+    school_lad = SchoolLocalAuthorityDistrict.find_or_initialize_by(
+      school: school,
+      local_authority_district: lad,
+    )
+
+    if school_lad.new_record?
+      school_lad.start_year = start_year
+      SchoolLocalAuthorityDistrict.find_by(school: school, end_year: nil)&.update!(end_year: start_year)
+    end
+
+    school_lad.save!
+  end
```

Comment:
I commonised this with some very rails-y metaprogramming, then hated it and deleted it immediately

```
def link_school_to_local_authority(school, local_authority)
  link_school_to_model(school, SchoolLocalAuthority, local_authority)
end

def link_school_to_lad(school, lad)
  link_school_to_model(school, SchoolLocalAuthorityDistrict, lad)
end

def link_school_to_model(school, model, instance)
  school_model = model.find_or_initialize_by(school: school, instance.class.to_s.underscore.to_sym => instance)

  if school_model.new_record?
    school_model.start_year = start_year
    model.find_by(school: school, end_year: nil)&.update!(end_year: start_year)
  end

  school_model.save!
end
```

Reply:
I think I hate this commonisation too.

Reply:
👍